servicemanager/servicemanager.go

```diff
@@ -632,29 +665,67 @@ func (sm *ServiceManager) extractPackageByURL(
 	}
 
 	if err = image.CheckFileInfo(context.Background(), sourceFile, fileInfo); err != nil {
-		return 0, aoserrors.Wrap(err)
+		return "", 0, nil, aoserrors.Wrap(err)
 	}
 
 	size, err := image.GetUncompressedTarContentSize(sourceFile)
 	if err != nil {
-		return 0, aoserrors.Wrap(err)
+		return "", 0, nil, aoserrors.Wrap(err)
 	}
 
-	if err = sm.tryAllocateSpace(size); err != nil {
+	if space, err = sm.serviceAllocator.AllocateSpace(uint64(size)); err != nil {
+		return "", 0, nil, aoserrors.Wrap(err)
+	}
+
+	imagePath, err = ioutil.TempDir(sm.servicesDir, "")
+	if err != nil {
+		return "", 0, nil, aoserrors.Wrap(err)
+	}
+
+	if err = image.UnpackTarImage(sourceFile, imagePath); err != nil {
+		return "", 0, nil, aoserrors.Wrap(err)
+	}
+
+	return imagePath, uint64(size), space, nil
+}
+
+func (sm *ServiceManager) getServiceSize(serviceID string) (size uint64, err error) {
+	services, err := sm.serviceInfoProvider.GetAllServiceVersions(serviceID)
+	if err != nil {
 		return 0, aoserrors.Wrap(err)
 	}
 
-	if err = image.UnpackTarImage(sourceFile, extractDir); err != nil {
-		return size, aoserrors.Wrap(err)
+	for _, service := range services {
+		size += service.Size
 	}
 
 	return size, nil
 }
 
-func (sm *ServiceManager) deallocateSpace(size int64) {
-	sm.Lock()
-	defer sm.Unlock()
+func acceptAllocatedSpace(spaceService *spaceallocator.Space, spacePackage *spaceallocator.Space) {
+	if err := spacePackage.Accept(); err != nil {
```

Comment:
I don't get clearly why we should accept both spaces. Do we really need both? Or there is an issue?

Reply:
@al1img  Yes. Because `prepareServiceFS` and `extractPackageByURL` use the same dir `servicesDir`.
I don't know why, but It's historically.